src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java

```diff
@@ -167,22 +224,196 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
+            throw new IllegalArgumentException("invalid format for min_score: must be a valid float value", e);
+        }
     }
 
     /**
      * Result class to hold min and max scores for each sub query
      */
     @AllArgsConstructor
     @Getter
-    private class MinMaxScores {
+    private static class MinMaxScores {
         float[] minScoresPerSubquery;
         float[] maxScoresPerSubquery;
     }
+
+    /**
+     * Result class to hold lower bound for each sub query
+     */
+    @Getter
+    static class LowerBound {
+        static final float MIN_LOWER_BOUND_SCORE = -10_000f;
+        static final float MAX_LOWER_BOUND_SCORE = 10_000f;
+        static final float DEFAULT_LOWER_BOUND_SCORE = 0.0f;
+
+        private final boolean enabled;
+        private final Mode mode;
+        private final float minScore;
+
+        LowerBound() {
+            this(false, Mode.DEFAULT, DEFAULT_LOWER_BOUND_SCORE);
+        }
+
+        LowerBound(boolean enabled, Mode mode, float minScore) {
+            this.enabled = enabled;
+            this.mode = mode;
+            this.minScore = minScore;
+        }
+
+        /**
+         * Enum for normalization mode
+         */
+        protected enum Mode {
+            APPLY {
+                @Override
+                public float normalize(float score, float minScore, float maxScore, float lowerBoundScore) {
+                    // if we apply the lower bound this mean we use actual score in case it's less then the lower bound min score
+                    // same applied to case when actual max_score is less than lower bound min score
+                    if (maxScore < lowerBoundScore || score < lowerBoundScore) {
+                        return (score - minScore) / (maxScore - minScore);
+                    }
+                    return (score - lowerBoundScore) / (maxScore - lowerBoundScore);
+                }
+            },
+            CLIP {
+                @Override
+                public float normalize(float score, float minScore, float maxScore, float lowerBoundScore) {
+                    // apply clipping, return lower bound min score if score is less than min score. This effectively means 0 after
+                    // normalization
+                    if (score < minScore) {
+                        return 0.0f;
+                    }
+                    if (maxScore < lowerBoundScore) {
+                        return (score - minScore) / (maxScore - minScore);
+                    }
+                    return (score - lowerBoundScore) / (maxScore - lowerBoundScore);
+                }
+            },
+            IGNORE {
+                @Override
+                public float normalize(float score, float minScore, float maxScore, float lowerBoundScore) {
+                    // ignore lower bound logic and do raw min-max normalization using actual scores
+                    float normalizedScore = (score - minScore) / (maxScore - minScore);
```

Comment:
How does the normalization logic handle negative scores? Are there edge cases when scores are negative, and have we added tests to cover these scenarios?

Reply:
this is "ignore" mode that is the same logic we do have today for min-max. in this case there is no lower bound score involved, so none of these scores should be negative because:
- all scores are positive
- min and max calculated based on those positive scores